### ISSUE TITLE
What is the point of the special node(s)

### ISSUE THREAD

**Issue by manojlds** [Samsung-AG/kraken#188]
following the code, I am trying to understand the point of special nodes.

Also, the `main.yaml` has tasks like:

```yaml
- include: node-pre.yaml
  when: "'specialnodes' in group_names"
```
but where is `specialnodes` defined as a group?

**Comment by maratoid** [comment]
Hi!

'specialnodes' as a group are defined in ansible inventory template:
https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/templates/ansible.inventory.tpl#L7-L8

In our AWS cluster we had a need to define some nodes that are not etcd, or kubernetes master/api server, but at the same time are are different from all the nodes in the autoscaling group. Default AWS configuration only has one special node - node-01, which needs to be a little beefier than everything in the ASG because we wanted it to make sure that some of the more resource-intensive services are tied to it. 

Basically specialnodes are one-off instances from bulk of the cluster brought up by the autoscaling group.

**Comment by manojlds** [comment]
Thanks for the explanation.  If my understanding is right, the `specialnodes` is not available as a group and the special node based conditions will all be false in the kubernetes role's `main.yaml`

**Comment by manojlds** [comment]
Talking about these: https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/tasks/main.yaml#L36

**Comment by maratoid** [comment]
If you look at https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/tasks/main.yaml#L34-L49 you'll see that all the same tasks are executed for sepcialnodes and nodes.

I should've mentioned that there are two ansible inventories rendered. 

One is created on each node type through cloud-init  - [here](https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/templates/node.yaml.tpl#L4-L34) is an example for a node. This inventory is then used by ansible docker container locally to the created node, like [so](https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/templates/node.yaml.tpl#L172-L184)

Another inventory is created locally on your machine, after the whole cluster is up - in case you want to run ansible on any/all nodes in the cluster, from [this](https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/templates/ansible.inventory.tpl) template. ASG ip addresses are appended to the rendered template by [kraken_asg_helper.sh](https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/kraken_asg_helper.sh#L77-L85) script.

It's a bit confusing, but it lets each node provision itself very quickly as it is coming up, and then lets you as a developer, make changes via ansible to all the nodes in bulk.
